### ISSUE TITLE
Pre-release versioning incompatible with [patch]

### ISSUE THREAD

**Issue by ryankurte** [sunng87/cargo-release#82]
Hey there,

I'm not sure whether this is strictly an _issue_, but it took me some time to work out and seems like a bit of a pain.

When `cargo release X` is run the version is bumped, the crate is published, then the version is again bumped to a development version. As cargo expects patch versions to match, this final bump means that you can no longer use `[patch]` overrides against the currently published library version for testing (without reverting the crate version in the patched crate's `Cargo.toml`).

This results in errors like:
```
warning: Patch `rr-mux v0.4.1-alpha.0 (/home/ryan/projects/rr-mux)` was not used in the crate graph.
```
Where the project `cargo.toml` contains:
```toml
[dependencies]
rr-mux = "0.4.0"
...

[patch.crates-io]
rr-mux = { path = "../rr-mux" }
```

One fix is to always run `cargo release` with `--no-dev-version` to avoid the problem. 

My suggestion would be that that the second development version bump should be _opt in_, I didn't expect it to happen, but, either way there's now some documentation for anyone else who has the same problem.

**Comment by sunng87** [comment]
The last bump behaviour was learned from tools in Java and Python world. It indicates the current branch may have changes after last release.

The `--no-dev-version` is created for your case. And you can add it to `release.toml` to make it a default option for your current project.

**Comment by ryankurte** [comment]
> The last bump behaviour was learned from tools in Java and Python world. It indicates the current branch may have changes after last release.

it _kindof_ makes sense in that regard, but it also seems incompatible with cargo. i personally would expect making a pre-release to also be an explicit command.

> The `--no-dev-version` is created for your case. And you can add it to `release.toml` to make it a default option for your current project.

yep, and i now use `--no-dev-version` every time. another possibility would be to add a global ~/.cargo/release.toml so it can be turned off completely.

as i said, i don't know whether to consider this _wrong_, it's just frustrating. i'm happy to open a PR altering the behaviour to be opt-in or documenting the incompatibility if one of those would be useful?

**Comment by sunng87** [comment]
> yep, and i now use `--no-dev-version` every time. another possibility would be to add a global `~/.cargo/release.toml` so it can be turned off completely. 

Yes you can add it to `~/.release.toml` (and I think `~/.cargo/release.toml` is a better place for this) to disable this behaviour globally. 

Also when the last version bump is disable, you will need to bump version manually when releasing a patch version. Because `cargo-release` will have no idea if current version is a released or unreleased one. So before I have a better solution to smooth this process, I suggest to keep current behaviour. 

Doc improvement for this is welcomed.

**Comment by epage** [comment]
Let me see if I understand the issue.  You have a repo with two crates, with one depending on the other.  You are declaring that dependency with a version and then using `[patch]` for local builds.

Is there a reason you are using `[patch]` rather than path dependencies?

From my understanding, path dependencies are intended for when you have multiple crates in a repo while `[patch]` is intended for if you need to override the version of a dependency with a different copy.  In my mind, you'd never publish anything in the `[patch]` table while you would publish path dependencies.

**Comment by ryankurte** [comment]
nope, i have two separate crates, one depending on the other, and patch is used to override this dependency with a local path so i can develop and test changes together.

**Comment by timfish** [comment]
I've come up against this again.

I know there is the `--no-dev-version` flag but this is not the default and most don't use it.

This result is that for the majority of repositories using `cargo-release`, `master` is on a new dev version which means that PR's cannot be tested with cargo patch. 

So in it's default configuration, `cargo-release` makes collaboration more cumbersome.

**Comment by epage** [comment]
We can probably look into updating patch.  The thing I[m trying to understand though (which affects priority) is why `[patch]` is being used, rather than path dependencies.

For example, I have the [`toml-test-rs`](https://github.com/epage/toml-test-rs) repo / workspace.  The [test harness has both version and path dependencies on other crates in the workspace](https://github.com/epage/toml-test-rs/blob/main/crates/harness/Cargo.toml).

**Comment by thomcc** [comment]
This is my only complaint about `cargo-release` (which I otherwise like), but it's really unfortunate, as it's slowly making `[patch]` less useful across the entire ecosystem :(.

That said, there seems to be confusion about what patch does, and what the issue is, how to fix it, ... so I'll try to clear it up. Sorry if it's a bit verbose.

> The thing I[m trying to understand though (which affects priority) is why `[patch]` is being used, rather than path dependencies.

`[patch]` is a tool for fully replacing some package everywhere it is used in the dependency graph. If the package only appears in the dependency graph once, then it's equivalent to adjusting the dependency line (e.g. switching `foo = "0.1.2"` to `foo = {path = ...}` somewhere).

However, that case isn't really the problem `[patch]` exists to solve, which is replacements in transitive dependencies (including dependencies used by many crates in my crate graph).

One important thing to note here: the version of the crate being replaced must exactly match the version of the crate that it's being replaced with. You cant use patch to change `foo:0.1.2` into `foo:0.1.3`, or `foo:0.1.3-alpha.0`. (After all, it's completely allowed for `foo:0.1.2` and `foo:0.1.3` to both exist in the dep tree at the same time)

This is why `cargo-release`'s default behavior causes issues — once the version number is bumped to `x.y.z-thing.n`, it can no longer be used patch an older version. This means that *immediately* after releasing, your crate cannot be patched (unless the user also undoes version bump, which isn't always viable).

---

Consider hitting a bug in something foundational, like `tokio` (or `serde`, although it's a bigger problem to have multiple tokios). and then wanting to see if the bug has already been fixed on their `master` branch.

With patch, you can just do this.
```
[patch.crates-io]
tokio = { git = "https://github.com/tokio-rs/tokio", branch = "master" }
```

But without `patch`, it becomes nearly impossible for big dependency trees — you'd need to fix it in every package in the dep tree which transitively depends on tokio.

While the current behavior of `cargo-release` doesn't break patch entirely, it does make it much harder to use, since you need to manually revert the change from dev bumping. In cases where a monorepo isn't in use (e.g. the `gfx-rs` ecosystem), this can spiral out of control, and become nearly impossible.

> We can probably look into updating patch

As might be clear by now the fix here isn't to update the versions in patch.

I think the only fix is make cargo-release not do the dev bump by default — It's a default with a very subtle and non-obvious downside, so in practice it probably will almost never get turned off.

**Comment by CAD97** [comment]
Minor nit @thomcc: crate@0.1.2 and crate@0.1.3 *can't* coëxist in the same dep graph; cargo prevents semver-compatible crate duplication. It's only if the versions are semver-incompatible which allows crate duplication. (This is the case even with more restrictive semver constraints than the default `^`.)

Otherwise, I'm in agreement: using prerelease versions for development is a cargo antipattern as of current, because cargo treats prereleases as semver incompatible with mainline releases, full stop.

If cargo treated upgrading to a semver compatible version with a prerelease string as semver compatible (but only by request), then this practice would be fine, and wouldn't break the use of `[patch]`.

I'm currently personally torn between "package.version is the in development version" and "package.version is the last released version". I'm currently on the side of "next version" (so the version number reflects what kind of changes are in tree), but my motivation here is thin.

I suspect what scheme you consider as best also depends on your exact branching workflow and release cadence.

**Comment by thomcc** [comment]
> Minor nit @thomcc: crate@0.1.2 and crate@0.1.3 can't coëxist in the same dep graph; cargo prevents semver-compatible crate duplication. It's only if the versions are semver-incompatible which allows crate duplication. (This is the case even with more restrictive semver constraints than the default ^.)

Ah, fair enough — 0.1.0 vs 0.2.0 then. Patch operates after versions are resolved so this doesn't change things really.

**Comment by thomcc** [comment]
This is bordering on off-topic, but I'm going to say it just to make it clear that the "next version" logic isn't a good alternative to the dev bump, and that really almost nobody should use it, unless they're very, *very* careful.

> I'm currently personally torn between "package.version is the in development version" and "package.version is the last released version". I'm currently on the side of "next version" (so the version number reflects what kind of changes are in tree), but my motivation here is thin.

I think "next version" is a bad idea for 99% of crates. It's extremely easy to get wrong (compared with "latest version" e.g. "only update the Cargo.toml version when publishing").

When updating the version for each change, the logic you need to use is subtle.

1. the contents of the change (if they're breaking)
2. the current package.version in Cargo.toml
3. the version number of the most recently released version on crates.io (without this, you'll skip numbers in releases.[^1])

But... Even this is not quite right. "most recently released version" is a value that changes over time, and so in development where you might have long-lived PRs and such, things can start right, become wrong, but get easy to miss, or get lost because two changes both try to do the same version bump.

The problem is that rule 3 here isn't clear — it says "most recently released version", but... most recent as of when? Unfortunately, it needs to be the most recent version as of merging the change into the repository. This seems like it would work perfectly, but is difficult to implement as it requires total control and knowledge of all changes, or an oracle that knows when your PRs will be done/merged.

So, why did I say 99% before and not 100%? Well, I guess there are cases where it doesn't hurt. Consider `tracing`: The changes for 0.2 are using something like the "next version" strategy. If something has a semver-major break, it gets the bump. Minor changes still use "latest version", and only works because it's a single version bump that is being considered, and there's a lot of process around changeing these versions. I think that applying this pattern in an unconsiered way would fail terribly.

[^1]: Consider: your crate is published as 0.1.0 in crates.io, and says the same in Cargo.toml's package.version in your repo.

    Over the next development cycle, you make 2 breaking changes, then 2 non-breaking changes. What should package.version in Cargo.toml say?

    If you don't consider what version is most recently released, you will get 0.3.2, but the correct answer is 0.2.0 (it was 0.1.0, and you have some breaking changes).

**Comment by epage** [comment]
Let me re-phrase to see if I understand.

The concern is not with "how do I develop my couple of crates in a workspace" but "project X doing process Y is making it harder for everyone else in the ecosystem, let's encourage friendlier processes by changing defaults"?

Breaking down
> project X doing process Y is making it harder for everyone else in the ecosystem

To apply this patch:
```
[patch.crates-io]
tokio = { git = "https://github.com/tokio-rs/tokio", branch = "master" }
```
tokio master has to be a compatible version with my dependency tree.  This means
- It can't be pre-release as thats always incompatible
- It can't be "next (incompatible) version" because that'll never work
- If they release an incompatible version, I have to find the last compatible ref that has the fix I need
  - **This** case makes me wonder if our `[[patch]]` workflow is busted.  Not being someone that uses it, I can't really say how this is in practice.  Can people clarify how much of a problem this is or isn't?

Another challenge with post-release version bump it makes it harder to auto-detect what crates have changed since the last release for #298 

So let's explore solving this:
- Change the default for post-release version bump
- Warn about `[patch]` on post-release version bump and use of prerelease
  - Maybe restrict this to crates without `[[bin]]`?
 
We still need to consider how to solve the problem this feature was intending to resolve.  If I understood @sunng87, the reason for the post-release-bump-to-prerelease is to help with support to identify the difference between a bug report from v0.1.0 and post-v0.1.0 `master`, is that right?  What if we instead encouraged programs to include their commit in the version by making it the default in `clap`?  I **think** a git commit is fine for the reproducibility folks.  As for `[[lib]]`s, unreleased crates are less likely to be used so this is less of a problem.

@Byron and @pksunkara, I'd be curious to hear your input on this.

**Comment by epage** [comment]
We've now released v0.18.0 with dev-version disabled by default.

Thank you all for your patience in explaining the impact and scope of this problem!

**Comment by pksunkara** [comment]
My bad for late reply, but yeah I would have also recommended disabling the `dev-version`.

> Some of them can instead be more general and come from maturity. In this case, my understanding is its a general release process problem that they were solving and not a tool / ecosystem problem. In contrast, it sounds like its as Rust tooling limitation that is preventing us from doing similar.

I was there 10 yrs when node.js ecosystem was still trying to figure out the same thing we are currently doing. I can confidently say that they reached maturity and decided that keeping the old version until the point of the release is the best way to deal with all kinds of dependency pathways (git, patch, etc..).

I disagree that it's a rust tooling limitation because I think it's a general dependency tooling limitation.
